packages/web-components/fast-foundation/src/carousel/carousel.ts

```diff
@@ -0,0 +1,411 @@
+            );
+        } else {
+            this.activeTabIndex = limit(
+                0,
+                this.tabs.length - 1,
+                this.activeTabIndex + adjustment
+            );
+        }
+        this.setComponent();
+    }
+
+    private slideIds: string[] = [];
+    private autoplayTimer: number | void;
+    private pausedTimeout: number | void;
+    private firstFocus: boolean = true;
+    private startTime: number = 0;
+    private stopTime: number = 0;
+
+    private incrementSlide = (direction: 1 | -1): void => {
+        const tempLength: number = this.basicPattern
+            ? this.items.length
+            : this.tabs.length;
+        const tempIndex: number = this.basicPattern
+            ? this.activeSlideIndex
+            : this.activeTabIndex;
+        this.focused = false;
+        let adjustment: number = 0;
+
+        if (this.basicPattern) {
+            if (this.loop) {
+                adjustment = wrapInBounds(0, tempLength - 1, tempIndex + direction);
+            } else {
+                adjustment = limit(0, tempLength - 1, tempIndex + direction);
+            }
+            this.activeSlideId = this.slideIds[adjustment];
+            this.itemsChanged();
+        } else {
+            this.adjust(direction);
+            this.focused = true;
+        }
+
+        this.change();
+    };
+
+    private autoplayNextItem = (): void => {
+        this.startTime = Date.now();
+        this.incrementSlide(1);
+    };
+
+    private generateSlideIds(): void {
+        this.slideIds = [];
+        this.slideIds = this.items.map((item: HTMLElement, index: number) => {
+            return item.getAttribute("id")
+                ? (item.getAttribute("id") as string)
+                : `slide-${index + 1}`;
+        });
+    }
+
+    private togglePlay(): void {
+        this.paused = !this.paused;
+    }
+
+    private startAutoPlay(): void {
+        if (!this.paused) {
+            this.stopAutoPlay();
+            this.startTime = Date.now();
+            this.autoplayTimer = window.setInterval(
+                this.autoplayNextItem,
+                this.autoplayInterval
+            );
+        }
+    }
+
+    private stopAutoPlay(): void {
+        this.autoplayTimer = window.clearInterval(this.autoplayTimer as number);
+    }
+
+    private handleRotationMouseDown = (e: Event): void => {
+        if (this.firstFocus) {
+            this.firstFocus = false;
+        }
+        this.togglePlay();
+    };
+
+    private handleRotationKeyDown = (e: KeyboardEvent) => {
+        switch (e.keyCode) {
+            case keyCodeEnter:
+            case keyCodeSpace:
+                this.handleRotationMouseDown(e);
+                break;
+        }
+    };
+
+    private handleFocusIn = (e: FocusEvent): void => {
+        // sethdonohue - per ARIA spec we need to stop rotation whenever keyboard focus is brought to the carousel unless the user specifically requests it to start again
```

Comment:
Can you put some these comments on new lines so they don't scroll the editor quite as much?